modules/ipc/include/hephaestus/ipc/zenoh/service.h

```diff
@@ -4,30 +4,187 @@
+    zenohc::QueryReplyOptions options;
+    if constexpr (std::is_same_v<ReplyT, std::string>) {
+      options.set_encoding(zenohc::Encoding{ Z_ENCODING_PREFIX_TEXT_PLAIN });  // Update encoding.
+      query.reply(this->topic_config_.name, reply, options);
+    } else {
+      options.set_encoding(zenohc::Encoding{ Z_ENCODING_PREFIX_EMPTY });  // Update encoding.
+      auto buffer = serdes::serialize(reply);
+      LOG(INFO) << "Replying with buffer of size: " << buffer.size();
+      query.reply(this->topic_config_.name, std::move(buffer), options);
+    }
+  };
+
+  queryable_ = expectAsUniquePtr(
+      session_->zenoh_session.declare_queryable(topic_config_.name, { std::move(query), []() {} }));
+}
+
+template <typename RequestT, typename ReplyT>
+auto callService(const SessionPtr& session, const TopicConfig& topic_config, const RequestT& request,
+                 const std::optional<std::chrono::milliseconds>& timeout /*= std::nullopt*/)
+    -> std::vector<ServiceResponse<ReplyT>> {
+  internal::checkTemplatedTypes<RequestT, ReplyT>();
+
+  LOG(INFO) << fmt::format("Calling service on '{}'", topic_config.name);
+  std::mutex m;
+  std::condition_variable done_signal;
+  std::vector<ServiceResponse<ReplyT>> reply_messages;
+
+  auto on_reply = [&reply_messages, &m, &topic_config](zenohc::Reply&& reply) {
+    const auto result = std::move(reply).get();
+    if (const auto& error = std::get_if<zenohc::ErrorMessage>(&result)) {
+      LOG(ERROR) << fmt::format("Error on '{}' reply: '{}'", topic_config.name, error->as_string_view());
+      return;
+    }
+
+    const auto& sample = std::get_if<zenohc::Sample>(&result);
+    LOG(INFO) << fmt::format("Received answer of '{}'", sample->get_keyexpr().as_string_view());
+    if constexpr (std::is_same_v<ReplyT, std::string>) {
+      CHECK(sample->get_encoding() == Z_ENCODING_PREFIX_TEXT_PLAIN)
+          << "Encoding for std::string should be Z_ENCODING_PREFIX_TEXT_PLAIN";
+      LOG(INFO) << fmt::format("Payload is string: '{}'", sample->get_payload().as_string_view());
+      std::unique_lock<std::mutex> lock(m);
+      reply_messages.emplace_back(ServiceResponse<ReplyT>{
+          .topic = topic_config.name,
+          .value = static_cast<std::string>(sample->get_payload().as_string_view()) });
+    } else {
+      CHECK(sample->get_encoding() == Z_ENCODING_PREFIX_EMPTY)
+          << "Encoding for binary types should be Z_ENCODING_PREFIX_EMPTY";
+      auto payload = sample->get_payload();
+      // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+      std::span<const std::byte> buffer(reinterpret_cast<const std::byte*>(payload.start), payload.len);
+      std::unique_lock<std::mutex> lock(m);
+      ReplyT reply_deserialized;
+      serdes::deserialize(buffer, reply_deserialized);
+      reply_messages.emplace_back(
+          ServiceResponse<ReplyT>{ .topic = topic_config.name, .value = std::move(reply_deserialized) });
```

Comment:
To get full benefit from the emplace
```suggestion
      reply_messages.emplace_back( topic_config.name, std::move(reply_deserialized));
```